### PR TITLE
Add Retired Key Management Slots

### PR DESCRIPTION
**Pull request by DaKnOb**
This Pull Request adds constants for the Retired Key Management slots of the YubiKey 5 series. This allows the possibility of using more key slots using this library.

### CODE REVIEW

**Comment by ericchiang** [comment]
Hey sorry for the delay here. Didn't have much bandwidth the last few weeks.

So a couple questions with the proposed API:
* It's kind of odd that this is a slice while the others are their own objects, do you expect users to do something like `piv.SlotRetiredKeyManagement[7]`?
* By definition these are retried and going to stop working in the future, what are the caveats here?
* You can already define these yourself. Maybe a comment would be better and projects can use them as they'd like to?

**Comment by DaKnOb** [comment]
It's okay, don't worry, we all do this on our personal free time, there's no expectations or a hurry :)

Indeed, I expect users to use them as you said, mainly because of the amount of them. Being able to iterate over them quickly and easily would help people use them more efficiently.
I think the "Retired" part of this does not mean they are no longer used. They are slots for retired keys. So if a GPG encryption key is rotated for example, the previous one is stored in one of these slots, and the new one in the appropriate slot. This is done so previous e-mails can still be decrypted successfully. This at least is according to my understanding of the documentation. In the Yubikey 5 Series, all these slots are available. Having the ability to access 20 more slots for certificates and keys, with a configurable touch policy and PIN policy for each one of them, in the same hardware device, allows for much greater flexibility for applications that want to use more keys, and it also allows normal "Key Retirement" operations using the library.

I have sent this patch to include this in the library as I believe it will be useful to more than just me. However, if you don't believe they should be included, this is totally fine :)

Thanks for the great work!

**Comment by ericchiang** [comment]
Ah okay, I misunderstood the "retirement" part of the slots. I thought it meant "depreciated" not "store old rotated keys here".

The currently the API is sharp because it's not clear what guarantees it would provide. Are we allow to change the order in the future? Or add more? If we ever add any we could break code that does something like:

```go
// We added a new slot and now this code returns a different value.
lastRetiredSlot := piv.SlotRetiredKeyManagement[len(piv.SlotRetiredKeyManagement)-1]
```

Or if we add new values, would looping over them potentially break old keys?

```go
for _, slot := range piv.SlotRetiredKeyManagement {
    // Woops, we added a new slot and now this loop accesses a slot
    // that the card doesn't support!
}
```

Maybe instead, something like this?

```go
// RetiredKeyManagementSlot provides access to "retired" slots. Slots meant for old Key Management
// keys that have been rotated. YubiKeys 4 and later support values between 0x82 and 0x95 (inclusive).
//
//     slot, ok := RetiredKeyManagementSlot(0x82)
//     if !ok {
//         // unrecognized slot
//     }
//     pub, err := yk.GenerateKey(managementKey, slot, key)
//
// https://developers.yubico.com/PIV/Introduction/Certificate_slots.html#_slot_82_95_retired_key_management
func RetiredKeyManagementSlot(key uint32) (Slot, bool) {
    // ...
}
```

So users have to declare specific slots: 

```go
slot, ok := RetiredKeyManagementSlot(0x82)
```

And can still loop over all the old slots, but in a way we don't break:

```go
var i uint32
for i = 0x82; i <= 0x95; i++ {
    slot, ok := RetiredKeyManagementSlot(i)
    // ...
}
```

**Comment by DaKnOb** [comment]
Indeed this seems like a better idea on how this could be implemented. I was also considering a `map`, however I thought since they were defined in the PIV standard, it was unlikely that they would change, or new ones would be added. I'll make the relevant modifications and let you know.

Thanks for the brainstorming

**Comment by DaKnOb** [comment]
Hello, I've made the necessary modifications and added tests for this, since it's now a function.

File: piv/key.go [hunk 1]
```diff
@@ -244,6 +244,48 @@ var (
...
+//
+//     slot, ok := RetiredKeyManagementSlot(0x82)
+//     if !ok {
+//         // unrecognized slot
+//     }
+//     pub, err := yk.GenerateKey(managementKey, slot, key)
+//
+// https://developers.yubico.com/PIV/Introduction/Certificate_slots.html#_slot_82_95_retired_key_management
+func RetiredKeyManagementSlot(key uint32) (Slot, bool) {
+	var SlotRetiredKeyManagement = []Slot{
```
Comment (ericchiang) [inline piv/key.go:258]: Define this globally, not within the function

File: piv/key.go [hunk 2]
```diff
@@ -244,6 +244,48 @@ var (
...
+		{0x90, 0x5fc11b},
+		{0x91, 0x5fc11c},
+		{0x92, 0x5fc11d},
+		{0x93, 0x5fc11e},
+		{0x94, 0x5fc11f},
+		{0x95, 0x5fc120},
+	}
+
+	if key >= 0x82 && key <= 0x95 {
+		return SlotRetiredKeyManagement[key-0x82], true
```
Comment (ericchiang) [inline piv/key.go:282]: Just use a map?

```go
val, ok := slotRetiredKeyManagement[key]
if ok {
    return Slot{key, val}, true
}
return Slot{}, false
```

File: piv/key_test.go [hunk 3]
```diff
@@ -579,3 +580,51 @@ func TestYubiKeyPrivateKeyPINError(t *testing.T) {
...
+			name:  "Non-existent slot, after range",
+			args:  args{key: 0x96},
+			want:  Slot{},
+			want1: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, got1 := RetiredKeyManagementSlot(tt.args.key)
+			if !reflect.DeepEqual(got, tt.want) {
```
Comment (ericchiang) [inline piv/key_test.go:622]: nit: types who's fields are comparable are themselves comparable, you don't need reflection and can just do:

```go
if got != tt.want || gotOK != tt.wantOK {
    t.Errorf(...)
}
```

File: piv/key_test.go [hunk 4]
```diff
@@ -579,3 +580,51 @@ func TestYubiKeyPrivateKeyPINError(t *testing.T) {
...
 	}
 }
+
+func TestRetiredKeyManagementSlot(t *testing.T) {
+	type args struct {
+		key uint32
+	}
+	tests := []struct {
+		name  string
+		args  args
```
Comment (ericchiang) [inline piv/key_test.go:590]: nit: just use "key uint32" instead of a struct here

File: piv/key_test.go [hunk 5]
```diff
@@ -579,3 +580,51 @@ func TestYubiKeyPrivateKeyPINError(t *testing.T) {
...
+		want  Slot
+		want1 bool
```
Comment (ericchiang) [inline piv/key_test.go:592]: nit: "wantOK" instead of "want1"

**Comment by DaKnOb** [comment]
The proposed changes have been implemented and pushed.

**Review by ericchiang — APPROVED** [review]
lgtm, please squash and I'll merge :)

**Comment by DaKnOb** [comment]
Done!

**Pull request merged**